### DISCUSSION TITLE
Create a `redefining` state in the registry

### DISCUSSION THREAD

**Discussion by hgrecco** [hgrecco/pint#1779] [Ideas]
As pointed out in #1773, when redefining a unit, certain caches needs to be invalidated. The easy way is to do it after each definition but this is expensive if it is used many times. One proposal would be to create a `defining` state to the registry to make this more performant.

In pseudo code would look something like this:
```python
class UnitRegistry:

    def define(self, ...):
        # actual code
        if not ureg._defining:
            self._build_cache()

    def defining():
        self._defining = True
        yield
        self._build_cache()
        self._defining = False

So single redefinitions can be added like now (but cache will be rebuild):
```python
ureg.define(...)
```

and when multiple are done
```python
with ureg.defining():
    ureg.define(...)
    ureg.define(...)
    # and many more
```

The other proposal would be to detect if a redefinition has occur. But we have to be careful about all alias, prefixed units and stuff. I like this second option more, but we need to be sure to make it right.

**Comment by hgrecco** [comment]
```python
    def load_definitions(self, ...):
        # actual code
        if not ureg._defining:
            self._build_cache()
```

**Comment by em1208** [comment]
@hgrecco The solution you proposed using `defining` looks fine to me but does it work also when using a new context and not just when adding a new definition?

**Comment by hgrecco** [comment]
Yes, the idea is that `defining` gets active when either `load_defintions` or `define` is called. And when it is then done, the cache is rebuilt.

**Reply by em1208** [reply]
Sounds good. How about when calling `enable_contexts`?

**Reply by em1208** [reply]
Hi @hgrecco, just checking, when do you plan to implement the proposal? Thanks!
